src/Databases/DatabaseFilesystem.cpp

```diff
@@ -0,0 +1,153 @@
+#include <Databases/DatabaseFilesystem.h>
+
+#include <IO/Operators.h>
+#include <IO/WriteBufferFromString.h>
+#include <Interpreters/Context.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ParserCreateQuery.h>
+#include <Parsers/parseQuery.h>
+#include <Storages/IStorage.h>
+#include <TableFunctions/TableFunctionFactory.h>
+
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_TABLE;
+}
+
+DatabaseFilesystem::DatabaseFilesystem(const String & name_, const String & path_, ContextPtr context_)
+    : IDatabase(name_), WithContext(context_->getGlobalContext()), path(path_), log(&Poco::Logger::get("DatabaseFileSystem(" + name_ + ")"))
+{
+    path = fs::path(path).lexically_normal().string();
+}
+
+std::string DatabaseFilesystem::getTablePath(const std::string & table_name) const
+{
+    return fs::path(path) / table_name;
+}
+
+void DatabaseFilesystem::addTable(const std::string & table_name, StoragePtr table_storage) const
+{
+    std::lock_guard lock(mutex);
+    loaded_tables.emplace(table_name, table_storage);
+}
+
+bool DatabaseFilesystem::isTableExist(const String & name, ContextPtr) const
+{
+    {
+        std::lock_guard lock(mutex);
+        if (loaded_tables.find(name) != loaded_tables.end())
+            return true;
+    }
+
+    fs::path table_file_path(getTablePath(name));
+    return fs::exists(table_file_path) && fs::is_regular_file(table_file_path);
+}
+
+StoragePtr DatabaseFilesystem::getTableImpl(const String & name, ContextPtr context_) const
+{
+    // Check if the table exists in the loaded tables map
+    {
+        std::lock_guard lock(mutex);
+        auto it = loaded_tables.find(name);
+        if (it != loaded_tables.end())
+            return it->second;
+    }
+
+    auto table_path = getTablePath(name);
+
+    // If the file exists, create a new table using TableFunctionFile and return it.
+    auto args = makeASTFunction("file", std::make_shared<ASTLiteral>(table_path));
+
+    auto table_function = TableFunctionFactory::instance().get(args, context_);
+    if (!table_function)
+        return nullptr;
+
+    auto table_storage = table_function->execute(args, context_, name);
+    if (table_storage)
+        addTable(name, table_storage);
+
+    return table_storage;
+}
+
+StoragePtr DatabaseFilesystem::getTable(const String & name, ContextPtr context_) const
+{
+    if (auto storage = getTableImpl(name, context_))
+        return storage;
+    throw Exception(ErrorCodes::UNKNOWN_TABLE, "Table {}.{} doesn't exist", backQuoteIfNeed(getDatabaseName()), backQuoteIfNeed(name));
+}
+
+StoragePtr DatabaseFilesystem::tryGetTable(const String & name, ContextPtr context_) const
+{
+    try
+    {
+        return getTable(name, context_);
+    }
+    catch (...)
+    {
+        return nullptr;
```

Comment:
Here `tryGetTable` means not that we ignore all exceptions from `getTable`, but that "return nullptr if table does not exist", any exceptions should be rethrown (at least other database engines work this way as well).

Reply:
I'm a little confused with exceptions handling here. The problem is that TableFunctionFile throws an exception if it failed to create table from file. We should turn such exceptions into `return nullptr;`. Because it means that the table does not exist, for example, BAD_ARGUMENTS it was not possible to understand the file extension. But it is unclear how to distinguish the expected exceptions from those where everything is completely broken, and should be rethrown.

It seems to me a good idea to catch all exceptions in `tryGetTable` and interpret them as  table does not exists. 
Now I'm trying to catch only those exceptions that I know are expected. But if someone adds new type of such exception, then tryGetTable will throw it.

Reply:
Firstly, at least a comment is needed like "Try creating StorageFile, which will throw an exception if file does not exist, we should catch this exception and return nullptr". Secondly, we could also check if file exists before creating the StorageFile object, thus return nullptr if std::filesystem::exists returns false, but as this does not protect from a race between checking for file existance and creating the StorageFile object - the try-catch is still needed. But in my opinion it is ok to catch only exception with error code `ErrorCodes::FILE_DOESNT_EXIST`, other exceptions can be rethrown. 

> But if someone adds new type of such exception, then tryGetTable will throw it.

this concern can be fixed when it arises, I think catching `ErrorCodes::FILE_DOESNT_EXIST` should still be enough.